pkg/agent/openflow/client.go

```diff
@@ -600,164 +600,124 @@ func (c *client) UninstallEndpointFlows(protocol binding.Protocol, endpoint prox
 		return fmt.Errorf("error when getting port: %w", err)
 	}
 	cacheKey := generateEndpointFlowCacheKey(endpoint.IP(), port, protocol)
-	return c.deleteFlows(c.serviceFlowCache, cacheKey)
+	return c.deleteFlows(c.featureService.cachedFlows, cacheKey)
 }
 
 func (c *client) InstallServiceFlows(groupID binding.GroupIDType, svcIP net.IP, svcPort uint16, protocol binding.Protocol, affinityTimeout uint16, nodeLocalExternal bool, svcType v1.ServiceType) error {
 	c.replayMutex.RLock()
 	defer c.replayMutex.RUnlock()
 	var flows []binding.Flow
-	flows = append(flows, c.serviceLBFlow(groupID, svcIP, svcPort, protocol, affinityTimeout != 0, nodeLocalExternal, svcType))
+	flows = append(flows, c.featureService.serviceLBFlow(groupID, svcIP, svcPort, protocol, affinityTimeout != 0, nodeLocalExternal, svcType))
 	if affinityTimeout != 0 {
-		flows = append(flows, c.serviceLearnFlow(groupID, svcIP, svcPort, protocol, affinityTimeout, nodeLocalExternal, svcType))
+		flows = append(flows, c.featureService.serviceLearnFlow(groupID, svcIP, svcPort, protocol, affinityTimeout, nodeLocalExternal, svcType))
 	}
 	cacheKey := generateServicePortFlowCacheKey(svcIP, svcPort, protocol)
-	return c.addFlows(c.serviceFlowCache, cacheKey, flows)
+	return c.addFlows(c.featureService.cachedFlows, cacheKey, flows)
 }
 
 func (c *client) UninstallServiceFlows(svcIP net.IP, svcPort uint16, protocol binding.Protocol) error {
 	c.replayMutex.RLock()
 	defer c.replayMutex.RUnlock()
 	cacheKey := generateServicePortFlowCacheKey(svcIP, svcPort, protocol)
-	return c.deleteFlows(c.serviceFlowCache, cacheKey)
+	return c.deleteFlows(c.featureService.cachedFlows, cacheKey)
 }
 
 func (c *client) GetServiceFlowKeys(svcIP net.IP, svcPort uint16, protocol binding.Protocol, endpoints []proxy.Endpoint) []string {
 	cacheKey := generateServicePortFlowCacheKey(svcIP, svcPort, protocol)
-	flowKeys := c.getFlowKeysFromCache(c.serviceFlowCache, cacheKey)
+	flowKeys := c.getFlowKeysFromCache(c.featureService.cachedFlows, cacheKey)
 	for _, ep := range endpoints {
 		epPort, _ := ep.Port()
 		cacheKey = generateEndpointFlowCacheKey(ep.IP(), epPort, protocol)
-		flowKeys = append(flowKeys, c.getFlowKeysFromCache(c.serviceFlowCache, cacheKey)...)
+		flowKeys = append(flowKeys, c.getFlowKeysFromCache(c.featureService.cachedFlows, cacheKey)...)
 	}
 	return flowKeys
 }
 
 func (c *client) InstallDefaultServiceFlows(nodePortAddressesIPv4, nodePortAddressesIPv6 []net.IP) error {
 	flows := []binding.Flow{
-		c.serviceNeedLBFlow(),
-		c.sessionAffinityReselectFlow(),
-		c.l2ForwardOutputServiceHairpinFlow(),
-	}
-	if c.networkConfig.IPv4Enabled {
-		flows = append(flows, c.serviceHairpinResponseDNATFlow(binding.ProtocolIP))
```

Comment:
Newly added  table `ServiceHairpinMarkTable` is used to mark hairpin traffic, and in table `SNATConntrackCommitTable`, hairpin traffic is performed SANT.

Reply:
The following function is used to bypass `-new+trk` hairpin packets.
```
// Feature: Service
// Stage: RoutingStage
// Tables: ServiceHairpinMarkTable
// hairpinBypassFlows generates the flows to classify hairpin connection and non-hairpin connection.
func (f *featureService) hairpinBypassFlows() []binding.Flow {
	cookieID := f.cookieAllocator.Request(f.category).Raw()
	var flows []binding.Flow
	for _, ipProtocol := range f.ipProtocols {
		flows = append(flows,
			// This generates the flow to mark the tracked hairpin connection with HairpinRegMark.
			ServiceHairpinMarkTable.ofTable.BuildFlow(priorityHigh).
				Cookie(cookieID).
				MatchProtocol(ipProtocol).
				MatchCTMark(HairpinCTMark).
				MatchCTStateNew(false).
				MatchCTStateTrk(true).
				Action().LoadRegMark(HairpinRegMark).
				Action().NextTable().
				Done(),
			// This generates the flow to bypass the non-hairpin connection.
			ServiceHairpinMarkTable.ofTable.BuildFlow(priorityNormal).
				Cookie(cookieID).
				MatchProtocol(ipProtocol).
				MatchCTStateNew(false).
				MatchCTStateTrk(true).
				Action().NextTable().
				Done(),
		)
	}
	return flows
}
```

The following two functions are used to mark `+new+trk` hairpin connections.
```
// Feature: Service
// Stage: RoutingStage
// Tables: ServiceHairpinMarkTable
// podHairpinFlow generates the flow to mark the first packet of hairpin connection from local Pods. Note that, value of
// ServiceSNATStateField will be overwritten to NotRequireSNATRegMark. The reason is that hairpin connection is forced to
// perform SNAT in SNATConntrackCommitTable regardless of whether the connection requires SNAT, and NotRequireSNATRegMark
// is one of match conditions to match the first packet of hairpin connection in SNATConntrackCommitTable. To make flow
// more simple in SNATConntrackCommitTable, NotRequireSNATRegMark will be loaded regardless of whether the connection
// requires SNAT. For this kind of hairpin connection, Antrea gateway IP is used to perform SNAT.
func (f *featureService) podHairpinFlow(endpoint net.IP) binding.Flow {
	ipProtocol := getIPProtocol(endpoint)
	return ServiceHairpinMarkTable.ofTable.BuildFlow(priorityLow).
		Cookie(f.cookieAllocator.Request(f.category).Raw()).
		MatchProtocol(ipProtocol).
		MatchCTStateNew(true).
		MatchCTStateTrk(true).
		MatchSrcIP(endpoint).
		MatchDstIP(endpoint).
		Action().LoadRegMark(HairpinRegMark).
		Action().LoadRegMark(SNATWithGatewayIP).
		Action().LoadRegMark(NotRequireSNATRegMark).
		Action().NextTable().
		Done()
}

// Feature: Service
// Stage: RoutingStage
// Tables: ServiceHairpinMarkTable
// gatewayHairpinFlows generate the flow to mark the first packet of hairpin Service connection from the Antrea gateway.
func (f *featureService) gatewayHairpinFlows() []binding.Flow {
	cookieID := f.cookieAllocator.Request(f.category).Raw()
	var flows []binding.Flow
	for _, ipProtocol := range f.ipProtocols {
		flows = append(flows,
			// Note that, in the flows generated by the following functions, value of ServiceSNATStateField will be overwritten
			// to NotRequireSNATRegMark. The reason is that hairpin connection is forced to perform SNAT in SNATConntrackCommitTable
			// regardless of whether the connection requires SNAT, and NotRequireSNATRegMark is one of match conditions to
			// match the first packet of hairpin connection in SNATConntrackCommitTable. To make flow more simple in
			// SNATConntrackCommitTable, NotRequireSNATRegMark will be loaded regardless of whether the connection requires
			// SNAT. For this kind of hairpin connection, virtual IP is used to perform SNAT since the packets destined
			// for the Antrea gateway.

			// This generates the flow to mark the first packet of Service connection sourced from the Antrea gateway and
			// destined for the Antrea gateway.
			ServiceHairpinMarkTable.ofTable.BuildFlow(priorityLow).
				Cookie(cookieID).
				MatchProtocol(ipProtocol).
				MatchCTStateNew(true).
				MatchCTStateTrk(true).
				MatchRegMark(GatewayHairpinRegMark).
				Action().LoadRegMark(HairpinRegMark).
				Action().LoadRegMark(NotRequireSNATRegMark).
				Action().LoadRegMark(SNATWithVirtualIP).
				Action().NextTable().
				Done(),
		)
	}
	return flows
}
```
Above three functions are used to mark hairpin

Reply:
I feel my understanding is messed up by the combined usage of GatewayHairpinRegMark, HairpinRegMark, NotRequireSNATRegMark, SNATWithVirtualIP, especially the 3rd mark. I know some comments try to explain the contradictory part, but it's still hard for me to associate all of these. I may need a sync-up whether this is really necessary to involve so many regs and keep the contradictory code.